### PR TITLE
Chat addition

### PR DESCRIPTION
**Pull request by comath**
Todo
- [x] Turn limits in the Challenge creation UI
- [x] Multiturn front end
- [x] Grading updates to support multi-turn
- [x] View updates to support multi-turn

### CODE REVIEW

File: llmv_models.py [hunk 1]
```diff
@@ -16,6 +17,22 @@
 
 log = getLogger(__name__)
 
+class LLMVChatPair(db.Model):
+    __tablename__ = 'llmv_chat_pair'
```
Comment (canyon289) [inline llmv_models.py:21]: What's pair mean in this context? Prompt response pair?

File: llmv_routes.py [hunk 2]
```diff
@@ -41,8 +42,44 @@ def generate_for_challenge():
...
+            if llmv_generation.account_id != get_current_user().id:
+                response = {'success': False, 'data': {'text': "This challenge is complete.", "id": -1}}
+                return jsonify(response)
+            if llmv_generation.challenge_id != challenge.id:
+                response = {'success': False, 'data': {'text': "This challenge is complete.", "id": -1}}
+                return jsonify(response)
+            history = llmv_generation.history
+            history = [h.json() for h in llmv_generation.history]
+            log.debug(f'Found history "{history}" ')
+        else:
```
Comment (canyon289) [inline llmv_routes.py:63]: This is the block that handles multi model setups?

File: llmv_routes.py [hunk 3]
```diff
@@ -133,6 +155,17 @@ def models_left(challenge_id):
         response = {'success': True, 'data': {"models_left": left_over_model}}
         return jsonify(response)
 
+    @llm_verifications.route('/chat_limit/<challenge_id>', methods=['GET'])
+    @authed_only
+    def chat_limit(challenge_id):
+        """Define a route for for showing users their answer submissions."""
```
Comment (canyon289) [inline llmv_routes.py:161]:
```suggestion
        """Define a route for showing users their answer submissions."""
```

File: migrations/5510e0688fa2_add_llmchallenge_tables.py [hunk 4]
```diff
@@ -37,8 +38,6 @@ def upgrade(op=None):
         sa.Column("model_id", sa.Integer(), nullable=False),
         sa.Column("user_id", sa.Integer(), nullable=True),
         sa.Column("team_id", sa.Integer(), nullable=True),
-        sa.Column("text", sa.Text(), nullable=False),
```
Comment (canyon289) [inline migrations/5510e0688fa2_add_llmchallenge_tables.py:40]: Moved to table that supports multiturn above if I understand correctly?

File: llmv_routes.py [hunk 5]
```diff
@@ -51,16 +88,10 @@ def generate_for_challenge():
         # Combine the pre-prompt and user-provided prompt with a space between them.
         prompt = request.json["prompt"]
         log.debug(f'pre-prompt {preprompt} and user-provided-prompt: "{prompt}"')
+        idempotency_uuid = str(uuid4())
```
Comment (canyon289) [inline llmv_routes.py:91]: Just curious why is this needed?

File: llmv_models.py [hunk 6]
```diff
@@ -16,6 +17,22 @@
...
+    __table_args__ = {'extend_existing': True}
+    id = db.Column(db.Integer, primary_key=True)
+    generation_id = db.Column(db.Integer, db.ForeignKey('llmv_generation.id', ondelete='CASCADE'))
+    prompt = db.Column(db.Text)
+    generation = db.Column(db.Text)
```
Comment (canyon289) [inline llmv_models.py:26]: Consider making these nullable in case the model fails for some reason and a generation is not returned

File: llmv_models.py [hunk 7]
```diff
@@ -16,6 +17,22 @@
 
 log = getLogger(__name__)
 
+class LLMVChatPair(db.Model):
+    __tablename__ = 'llmv_chat_pair'
+    __table_args__ = {'extend_existing': True}
```
Comment (canyon289) [inline llmv_models.py:22]: Is this table defined elsewhere?

https://docs.sqlalchemy.org/en/13/core/metadata.html#sqlalchemy.schema.Table.params.extend_existing

File: llmv_routes.py [hunk 8]
```diff
@@ -41,8 +42,44 @@ def generate_for_challenge():
         """Add a route to CTFd for generating text from a prompt."""
         log.info(f'Received text generation request from user "{get_current_user().name}" '
                  f'for challenge ID "{request.json["challenge_id"]}"')
+        
         challenge = LlmChallenge.query.filter_by(id=request.json['challenge_id']).first_or_404()
 
+        if 'generation_id' in request.json:
+            log.info("Found old generation id, using that")
```
Comment (canyon289) [inline llmv_routes.py:49]:
```suggestion
            log.info("Found old generation id %s, using that", request.json['generation_id'])
```

Comment (canyon289) [reply llmv_routes.py:49]: Will make debugging a bit easier

File: llmv_routes.py [hunk 9]
```diff
@@ -41,8 +42,44 @@ def generate_for_challenge():
...
+            llmv_generation = LLMVGeneration.query.filter_by(id=request.json['generation_id']).first_or_404()
+            if llmv_generation.status != "pending":
```
Comment (canyon289) [inline llmv_routes.py:51]: Style nit: All these conditions return the same response. Consider collapsing unless the plan is to change the returned json
```
            if (llmv_generation.status != "pending" or 
               llmv_generation.account_id != get_current_user().id or 
               llmv_generation.challenge_id != challenge.id):

               response = {'success': False, 'data': {'text': "This challenge is complete.", "id": -1}}
               return jsonify(response)

File: llmv_routes.py [hunk 10]
```diff
@@ -41,8 +42,44 @@ def generate_for_challenge():
...
+                return jsonify(response)
+            if llmv_generation.account_id != get_current_user().id:
+                response = {'success': False, 'data': {'text': "This challenge is complete.", "id": -1}}
+                return jsonify(response)
+            if llmv_generation.challenge_id != challenge.id:
+                response = {'success': False, 'data': {'text': "This challenge is complete.", "id": -1}}
+                return jsonify(response)
+            history = llmv_generation.history
+            history = [h.json() for h in llmv_generation.history]
+            log.debug(f'Found history "{history}" ')
```
Comment (canyon289) [inline llmv_routes.py:62]: Very optional suggestion, lazy evaluation is suggested for logging statements

```suggestion
            log.debug('Found history %s ', history)
```

File: migrations/5510e0688fa2_add_llmchallenge_tables.py [hunk 11]
```diff
@@ -49,6 +48,16 @@ def upgrade(op=None):
         sa.ForeignKeyConstraint(["user_id"], ["users.id"]),
         sa.PrimaryKeyConstraint("id"),
     )
+    op.create_table(
```
Comment (canyon289) [inline migrations/5510e0688fa2_add_llmchallenge_tables.py:51]: Its good to see migrations. Will a contest ever be migrated though? It feels like a new db would be started each time an event is run meaning no migration would need to take place?

Cant hurt to have it, just asking

File: remote_llm.py [hunk 12]
```diff
@@ -3,14 +3,16 @@
...
 
 # Third-party imports.
-from requests import post, get
+import requests
 from requests.exceptions import HTTPError
 
 log = getLogger(__name__)
 
-def generate_text(preprompt, prompt, model):
+def generate_text(idempotency_uuid, preprompt, prompt, model, history= []):
```
Comment (canyon289) [inline remote_llm.py:15]: Dont use mutables in python arguments 

https://docs.python-guide.org/writing/gotchas/
```suggestion
def generate_text(idempotency_uuid, preprompt, prompt, model, history= None):
  if history is None:
    history = []
```

File: remote_llm.py [hunk 13]
```diff
@@ -34,10 +36,15 @@ def generate_text(preprompt, prompt, model):
...
     # Load the Vanilla Neox API key from the config file.
-
-    raw_response = post(url=route,
-                        headers={'Authorization': f'Bearer {token}'},
-                        json={'prompt': prompt, "preprompt" : preprompt, "model": model})
+    try:
+        raw_response = requests.post(url=route,
+                            headers={'Authorization': f'Bearer {token}'},
+                            json={"uuid": idempotency_uuid,'prompt': prompt, "system" : preprompt, "model": model, "history": history})
+    except requests.Timeout:
```
Comment (canyon289) [inline remote_llm.py:43]: This should work too. Adding logging for extra debugging
```suggestion
    except (requests.Timeout, requests.ConnectionError) as error:
      logging.debug(error)
```

**Comment by canyon289** [comment]
@comath I fixed teh conflict

File: llmv_routes.py (hunk 9 above)
Comment (comath) [reply llmv_routes.py:51]: It should have custom error messages.

File: remote_llm.py (hunk 12 above)
Comment (comath) [reply remote_llm.py:15]: Python is a foreign land!

File: migrations/5510e0688fa2_add_llmchallenge_tables.py (hunk 11 above)
Comment (comath) [reply migrations/5510e0688fa2_add_llmchallenge_tables.py:51]: This is where the tables are created, the `llm_model` file just reads the established tables.

File: assets/view.js [hunk 14]
```diff
@@ -60,33 +66,72 @@ Alpine.data("llm_verification", () => ({
...
   async init() {
+    url = CTFd.config.urlRoot + `/chat_limit/` + this.id;
+
+    const response = await CTFd.fetch(url, {
+      method: "get",
+    });
+    const result = await response.json();
+    this.chat_limit = result.data.chat_limit;
+    if (this.chat_limit > 1) {
+      this.is_single_turn = false;
```
Comment (canyon289) [inline assets/view.js:77]: Why two bools? Cant single turn vs multi turn be differentiated with just one?

File: remote_llm.py [hunk 15]
```diff
@@ -33,11 +37,17 @@ def generate_text(preprompt, prompt, model):
...
-    raw_response = post(url=route,
-                        headers={'Authorization': f'Bearer {token}'},
-                        json={'prompt': prompt, "preprompt" : preprompt, "model": model})
+    try:
+        raw_response = requests.post(url=route,
+                            headers={'Authorization': f'Bearer {token}'},
+                            json={"uuid": idempotency_uuid,'prompt': prompt, "system" : preprompt, "model": model, "history": history})
+    except (requests.Timeout, requests.ConnectionError) as error:
+        log.debug(error)
+        # try again, with idempotency key
```
Comment (canyon289) [inline remote_llm.py:47]: Should something happen here? Connection error is just handled twice?

**Comment by canyon289** [comment]
From static read LGTM, Merge it and fixups can happen on main

**Comment by canyon289** [comment]
Merging so I can continue fixing things up on main
